header/p2p/server.go

```diff
@@ -52,19 +54,31 @@ func (serv *ExchangeServer) Stop(context.Context) error {
 
 // requestHandler handles inbound ExtendedHeaderRequests.
 func (serv *ExchangeServer) requestHandler(stream network.Stream) {
+	err := stream.SetReadDeadline(time.Now().Add(readDeadline))
+	if err != nil {
+		log.Debugf("error setting deadline: %s", err)
+	}
 	// unmarshal request
 	pbreq := new(p2p_pb.ExtendedHeaderRequest)
-	_, err := serde.Read(stream, pbreq)
+	_, err = serde.Read(stream, pbreq)
 	if err != nil {
 		log.Errorw("server: reading header request from stream", "err", err)
 		stream.Reset() //nolint:errcheck
 		return
 	}
+	if err = stream.CloseRead(); err != nil {
+		log.Warn(err)
+	}
 	// retrieve and write ExtendedHeaders
-	if pbreq.Hash != nil {
-		serv.handleRequestByHash(pbreq.Hash, stream)
-	} else {
-		serv.handleRequest(pbreq.Origin, pbreq.Origin+pbreq.Amount, stream)
+	switch pbreq.Data.(type) {
+	case *p2p_pb.ExtendedHeaderRequest_Hash:
+		serv.handleRequestByHash(pbreq.GetHash(), stream)
+	case *p2p_pb.ExtendedHeaderRequest_Origin:
+		serv.handleRequest(pbreq.GetOrigin(), pbreq.GetOrigin()+pbreq.Amount, stream)
+	default:
+		log.Error("server: invalid data type received")
```

Comment:
We should eventually have some sort of blocking behaviour for peers that send us multiple bad requests. cc @Wondertan 

This should be p2p_exchange - wide IMO where we track peer IDs who send us bad requests server-side and blocklist them if they hit 2+ bad reqs. WDYT? We can stick this in the epic as an issue. I thought we had some notes about that in the `Head` sanity checking issue but we should have it on the server side for any inbound request that is bad.

Reply:
There should be some single peer score tracking component that each protocol we have can report to. Sending bad data is something every protocol faces. What we could do for now, is just to blacklist a peer completely.

Reply:
Yeah, let's flesh this out post-mainnet. Maybe create a stub issue for this in the moment? Doesn't have to be fleshed out fully.